workers/limiter/src/index.ts

```diff
@@ -0,0 +1,310 @@
+import { DatabaseController } from '../../../lib/db/controller';
+import { Worker } from '../../../lib/worker';
+import * as pkg from '../package.json';
+import asyncForEach from '../../../lib/utils/asyncForEach';
+import { Collection, Db } from 'mongodb';
+import * as path from 'path';
+import * as dotenv from 'dotenv';
+import { PlanDBScheme, ProjectDBScheme, WorkspaceDBScheme } from 'hawk.types';
+import redis from 'redis';
+import HawkCatcher from '@hawk.so/nodejs';
+import axios from 'axios';
+import shortNumber from 'short-number';
+import ReportData from '../types/reportData';
+
+/**
+ * Workspace with its tariff plan
+ */
+type WorkspaceWithTariffPlan = WorkspaceDBScheme & {tariffPlan: PlanDBScheme}
+
+dotenv.config({ path: path.resolve(__dirname, '../.env') });
+
+/**
+ * Worker for checking current total events count in workspaces and limits events receiving if workspace exceed the limit
+ */
+export default class LimiterWorker extends Worker {
+  /**
+   * Worker type
+   */
+  public readonly type: string = pkg.workerType;
+
+  /**
+   * Database Controller for events database
+   */
+  private eventsDb: DatabaseController = new DatabaseController(process.env.MONGO_EVENTS_DATABASE_URI);
+
+  /**
+   * Database Controller for accounts database
+   */
+  private accountsDb: DatabaseController = new DatabaseController(process.env.MONGO_ACCOUNTS_DATABASE_URI);
+
+  /**
+   * Connection to events DB
+   */
+  private eventsDbConnection!: Db;
+
+  /**
+   * Collection with projects
+   */
+  private projectsCollection!: Collection<ProjectDBScheme>;
+
+  /**
+   * Collection with workspaces
+   */
+  private workspacesCollection!: Collection<WorkspaceDBScheme>;
+
+  /**
+   * Redis client for making queries
+   */
+  private readonly redisClient = redis.createClient({ url: process.env.REDIS_URL });
+
+  /**
+   * Redis key for storing banned projects
+   */
+  private readonly redisDisabledProjectsKey = 'DisabledProjectsSet'
+
+  /**
+   * Start consuming messages
+   */
+  public async start(): Promise<void> {
+    this.eventsDbConnection = await this.eventsDb.connect();
+    const accountDbConnection = await this.accountsDb.connect();
+
+    this.projectsCollection = accountDbConnection.collection<ProjectDBScheme>('projects');
+    this.workspacesCollection = accountDbConnection.collection<WorkspaceDBScheme>('workspaces');
+    await super.start();
+  }
+
+  /**
+   * Finish everything
+   */
+  public async finish(): Promise<void> {
+    await super.finish();
+    await this.eventsDb.close();
+    await this.accountsDb.close();
+  }
+
+  /**
+   * Task handling function
+   */
+  public async handle(): Promise<void> {
+    this.logger.info('Limiter worker started task');
+    const reportData: ReportData = {
+      workspacesWithoutLastChargeDate: new Set(),
+      bannedWorkspaces: [],
+      bannedProjectIds: [],
+    };
+    const projectIdsToBan = await this.getProjectsIdsToBan(reportData);
```

Comment:
зачем во все методы прикидывается `reportData`? не красиво как-то. Они могут возвращать просто данные наверх? и в handle собрать уже reportData

Reply:
тогда будут некрасивые возвращаемые значения(
либо объект с дополнительной информацией возвращать надо будет, либо массив

Reply:
нормально будет, порефактори только немного 

```suggestion
    const projectToBan = await this.getProjectsToBan();
```

и тд

Reply:
ну а инфу о забаненных воркспейсах о воркспейсах без даты последнего списания как прокидывать?
возвращать вместе с проектами из getProjectsToBan? и их других функций тоже?

Reply:
нет, надо декомпозировать `getProjectsIdsToBan` на методы. Сейчас он много всего делает, даже апдейты какие-то — это не соответствует названию и отсюда у тебя проблемы с прокидыванием

Reply:
и наверное можно переделать на getWorkspacesToBan , которая вернет воркспейсы и проекты